fdbserver/SimulatedCluster.actor.cpp

```diff
@@ -2520,31 +2543,31 @@ ACTOR void setupAndRun(std::string dataFolder,
 	state Optional<TenantName> defaultTenant;
 	state Standalone<VectorRef<TenantNameRef>> tenantsToCreate;
 	state TenantMode tenantMode = TenantMode::DISABLED;
-	if (tenantModeRequired || (allowDefaultTenant && deterministicRandom()->random01() < 0.5)) {
-		defaultTenant = "SimulatedDefaultTenant"_sr;
-		tenantsToCreate.push_back_deep(tenantsToCreate.arena(), defaultTenant.get());
-		if (tenantModeRequired || deterministicRandom()->random01() < 0.9) {
-			tenantMode = TenantMode::REQUIRED;
+	// If this is a restarting test, restartInfo.ini is read in restartSimulatedSystem
+	// where we update the defaultTenant and tenantMode in the testConfig
+	// Defer setting tenant mode and default tenant until later
+	if (!rebooting) {
+		if (testConfig.tenantModes.size()) {
+			auto randomPick = deterministicRandom()->randomChoice(testConfig.tenantModes);
+			tenantMode = TenantMode::fromValue(StringRef(std::to_string(randomPick)));
+			if (tenantMode == TenantMode::REQUIRED) {
+				allowDefaultTenant = true;
```

Comment:
This is just trying to emulate the previous behaviour where the only way `tenantMode = required` could happen was if the default tenant was set. Was that not done intentionally before this change?

Reply:
The logic for this is inverted from the previous case. Before, we would choose whether to use a default tenant and then sometimes randomly set tenant mode to required if we are using one, but now we are choosing the tenant mode first and then deciding what to do for the default tenant. In the first case, we would not want to enable required mode for a test that hasn't opted in to it and may not be using tenants outside of the default tenant.

On the other hand, a test that specifies that tenants are required does not necessarily also want to use a default tenant (it could be managing tenants itself), and if the default tenant is explicitly disabled we should not set one. If `allowDefaultTenant` is true, though, I could see the argument for always using it.

Reply:
I see, I'll make that change then